sources/sql_database/pg_replication/helpers.py

```diff
@@ -0,0 +1,528 @@
+    qual_name = _make_qualified_table_name(table_name, schema_name)
+    esc_pub_name = escape_postgres_identifier(publication_name)
+    try:
+        cur.execute(f"ALTER PUBLICATION {esc_pub_name} ADD TABLE {qual_name};")
+        logger.info(
+            f"Successfully added table {qual_name} to publication {esc_pub_name}."
+        )
+    except psycopg2.errors.DuplicateObject:  # table is already member of publication
+        pass
+
+
+def create_replication_slot(  # type: ignore[return]
+    name: str, cur: ReplicationCursor, output_plugin: str = "pgoutput"
+) -> Optional[Dict[str, str]]:
+    """Creates a replication slot if it doesn't exist yet."""
+    try:
+        cur.create_replication_slot(name, output_plugin=output_plugin)
+        logger.info(f'Successfully created replication slot "{name}".')
+        result = cur.fetchone()
+        return {
+            "slot_name": result[0],
+            "consistent_point": result[1],
+            "snapshot_name": result[2],
+            "output_plugin": result[3],
+        }
+    except psycopg2.errors.DuplicateObject:  # the replication slot already exists
+        logger.info(
+            f'Replication slot "{name}" cannot be created because it already exists.'
+        )
+
+
+def drop_replication_slot(name: str, cur: ReplicationCursor) -> None:
+    """Drops a replication slot if it exists."""
+    try:
+        cur.drop_replication_slot(name)
+        logger.warning(f'Successfully dropped replication slot "{name}".')
+    except psycopg2.errors.UndefinedObject:  # the replication slot does not exist
+        logger.warning(
+            f'Replication slot "{name}" cannot be dropped because it does not exist.'
+        )
+
+
+def drop_publication(name: str, cur: ReplicationCursor) -> None:
+    """Drops a publication if it exists."""
+    esc_name = escape_postgres_identifier(name)
+    try:
+        cur.execute(f"DROP PUBLICATION {esc_name};")
+        cur.connection.commit()
+        logger.warning(f"Successfully dropped publication {esc_name}.")
+    except psycopg2.errors.UndefinedObject:  # the publication does not exist
+        logger.warning(
+            f"Publication {esc_name} cannot be dropped because it does not exist."
+        )
+
+
+def persist_snapshot_table(
+    snapshot_name: str,
+    table: str,
+    schema: str,
+    cur: psycopg2.extensions.cursor,
+    include_columns: Optional[Sequence[str]] = None,
+) -> str:
+    """Persists exported snapshot table state.
+
+    Reads snapshot table content and copies it into new table.
+    """
+    col_str = "*"
+    if include_columns is not None:
+        col_str = ", ".join(map(escape_postgres_identifier, include_columns))
+    snapshot_table_name = f"{table}_snapshot_{snapshot_name}"
+    snapshot_qual_name = _make_qualified_table_name(snapshot_table_name, schema)
+    qual_name = _make_qualified_table_name(table, schema)
+    cur.execute(
+        f"""
+        START TRANSACTION ISOLATION LEVEL REPEATABLE READ;
```

Comment:
why do we need repeatable read?

Reply:
Doesn't work with `READ COMMITTED`:
![image](https://github.com/dlt-hub/verified-sources/assets/47451109/bd4afba2-488e-4507-aae0-6839c0656a20)

Also mentioned on SO: https://stackoverflow.com/a/76284206